### Homework 3 Numpy Arrays and Pandas

1. Open `cars.csv` in jupyter and look at the data. (Double click `cars.csv` to open in a new tab).
2. We are converting each **column** into a numpy array.
3. Notice we have text, integer and float data. 
4. Take care not to overwrite your starter code.

In [6]:
#Starter code. Do NOT modify this cell
import pandas as pd
import numpy as np

# Read in dataset
df = pd.read_csv('cars.csv',delimiter=';')

# Convert each column to an np.array
car = df.Car.to_numpy()
mpg = df.MPG.to_numpy()
cylinders = df.Cylinders.to_numpy()
displacement = df.Displacement.to_numpy()
horsepower = df.Horsepower.to_numpy()
weight = df.Weight.to_numpy()
acceleration = df.Acceleration.to_numpy()
model = df.Model.to_numpy()
origin = df.Origin.to_numpy()

# create an array of headers
headers = np.array(['Car', 'MPG', 'Cylinders', 'Displacement', 'Horsepower', 'Weight','Acceleration', 'Model', 'Origin'])


1. (2 pts) What are the numpy datatypes for the `car`,`mpg`, `horsepower` and `origin` arrays? (Hint: Not `type(car)` and use `print()` to print the numpy type.)

In [13]:
arr_1 = np.array(['Car', 'MPG', 'Horsepower', 'Origin'])
print('Array:', arr_1)
print('Datatype:', arr_1.dtype)

Array: ['Car' 'MPG' 'Horsepower' 'Origin']
Datatype: <U10


2. (1 pt) `numpy` is often imported as `np`. What is this action called? Why might you do this? 

This allows for a more effective way of writing code as np is a shortened form of numpy and allows an easier way of writing the code.

3. (2 pts) Each of the arrays (cars, mpg, cylinders, etc..) has how many dimensions? Show how to find this using code. 

In [14]:
a = np.array(['Car', 'MPG', 'Cylinders', 'Displacement', 'Horsepower', 'Weight', 'Acceleration', 'Model', 'Origin'])
print(a.ndim)

1


In [15]:
b = np.array(['Car'])
print(a.ndim)

1


In [ ]:
# Each of the arrays has one dimension

4. (3 pts) What is the ___ of `horsepower`?
- sum
- min
- max
- median
- Optional: mode (May need to google help) (1 pt extra credit)

In [16]:
df.head()

Car   MPG  Cylinders  Displacement  Horsepower  \
0  Chevrolet Chevelle Malibu  18.0          8         307.0       130.0   
1          Buick Skylark 320  15.0          8         350.0       165.0   
2         Plymouth Satellite  18.0          8         318.0       150.0   
3              AMC Rebel SST  16.0          8         304.0       150.0   
4                Ford Torino  17.0          8         302.0       140.0   

   Weight  Acceleration  Model Origin  
0  3504.0          12.0     70     US  
1  3693.0          11.5     70     US  
2  3436.0          11.0     70     US  
3  3433.0          12.0     70     US  
4  3449.0          10.5     70     US

In [27]:
df["Horsepower"].sum()

42033.0

In [28]:
df['Horsepower'].min()

0.0

In [29]:
df['Horsepower'].max()

230.0

In [30]:
df["Horsepower"].median()

93.5

In [31]:
df["Horsepower"].mode()

0    150.0
Name: Horsepower, dtype: float64

5. (2 pts) Select the last element in each subarray (do this with one operation.)

In [32]:
select_more = np.array([[1,2,3],
                      [4,5,6],
                      [7,8,9]])
# returns array([3, 6, 9])

In [36]:
def Extract(select_more):
    return [item[-1] for item in select_more]
print(Extract(select_more))

[3, 6, 9]


6. (2 pts) Use the built-in `np.lookfor()` function to search for `'log'`.

Use the function you find and apply it to `mpg`.

In [50]:
a = df["MPG"].mean()
np.log(a)

3.13771919652272

7. (2 pts) Use boolean indexing to create an array of `True/False` using `origin` and the value 'US'.

In [31]:
print(np.array)

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True, False,  True,  True,  True,  True,  True,  True,  True,
        True,  True, False,  True,  True,  True, False, False, False,
       False, False, False,  True,  True,  True,  True,  True, False,
        True, False,  True, False,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True, False, False, False, False, False, False,
        True, False,  True, False,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True, False,  True,  True,
        True,  True, False, False, False, False,  True, False, False,
        True, False,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True, False,  True,  True,  True,  True,  True, False,  True,
       False, False,  True,  True, False,  True,  True, False, False,
       False, False,

In [25]:
df.head(10)

Car   MPG  Cylinders  Displacement  Horsepower  \
0  Chevrolet Chevelle Malibu  18.0          8         307.0       130.0   
1          Buick Skylark 320  15.0          8         350.0       165.0   
2         Plymouth Satellite  18.0          8         318.0       150.0   
3              AMC Rebel SST  16.0          8         304.0       150.0   
4                Ford Torino  17.0          8         302.0       140.0   
5           Ford Galaxie 500  15.0          8         429.0       198.0   
6           Chevrolet Impala  14.0          8         454.0       220.0   
7          Plymouth Fury iii  14.0          8         440.0       215.0   
8           Pontiac Catalina  14.0          8         455.0       225.0   
9         AMC Ambassador DPL  15.0          8         390.0       190.0   

   Weight  Acceleration  Model Origin  
0  3504.0          12.0     70     US  
1  3693.0          11.5     70     US  
2  3436.0          11.0     70     US  
3  3433.0          12.0     70     US  
4  3449.0          10.5     70     US  
5  4341.0          10.0     70     US  
6  4354.0           9.0     70     US  
7  4312.0           8.5     70     US  
8  4425.0          10.0     70     US  
9  3850.0           8.5     70     US

In [35]:
dict = df
a = pd.DataFrame(dict, index = [False, True, True, False])
print(a)

       Car  MPG  Cylinders  Displacement  Horsepower  Weight  Acceleration  \
False  NaN  NaN        NaN           NaN         NaN     NaN           NaN   
True   NaN  NaN        NaN           NaN         NaN     NaN           NaN   
True   NaN  NaN        NaN           NaN         NaN     NaN           NaN   
False  NaN  NaN        NaN           NaN         NaN     NaN           NaN   

       Model Origin  
False    NaN    NaN  
True     NaN    NaN  
True     NaN    NaN  
False    NaN    NaN  


8. (2 pts) Apply your index (from above) to the `car` array. You should now only see cars that are of US origin. 

In [21]:
df.iloc[:, 8].head()

0    US
1    US
2    US
3    US
4    US
Name: Origin, dtype: object

### Use numpy or pandas to answer the following questions.

9. (3 pts) What is the average horsepower for `Europe` cars? (Hint: create a boolean index of 'US' apply it to horsepower and perform the calculation.)

In [23]:
grouped_df = df.groupby('Origin')
print(grouped_df)

In [24]:
grouped_df['Horsepower'].mean()

Origin
Europe     78.780822
Japan      79.835443
US        118.011811
Name: Horsepower, dtype: float64

The average horsepower for Europe cars is 78.780822 HP.

10. (3 pts) On average are Japanese cars more fuel-efficient than US cars? Show the code that you used to determine this.  

In [51]:
grouped_df = df.groupby('Origin')
print(grouped_df)

In [53]:
grouped_df['MPG'].max()

Origin
Europe    44.3
Japan     46.6
US        39.0
Name: MPG, dtype: float64

Japanese cars are more fuel efficient as they tend to have a higer MPG resulting in a higher fuel efficiency.

11. (3 pts) What is the correlation coefficient of `mpg` and `weight`? What is a correlation coefficient? What does the result indicate about the relationship between mpg and weight? 

In [47]:
np.corrcoef(df['MPG'],df['Weight'])

array([[ 1.        , -0.78750994],
       [-0.78750994,  1.        ]])

Correlation coefficient indicates how strong a relationship is between two variables. In this case, there appears to be a high negative correlation between mpg and weight.

Extra credit: (3 pts) Create a new pandas dataframe that has the average mpg, weight, and horsepower per country of origin

In [48]:
my_list = [{"country_of_origin": "Canada", "mpg": 31, "weight": 3337, "horsepower": 150}, 
           {"country_of_origin": "Germany", "mpg": 35, "weight": 3361, "horsepower": 255}, 
           {"country_of_origin": "SK", "mpg": 25, "weight": 3396, "horsepower": 187}]
my_df_1 = pd.DataFrame(my_list)
my_df_1

country_of_origin  mpg  weight  horsepower
0            Canada   31    3337         150
1           Germany   35    3361         255
2                SK   25    3396         187